systemd/datadog_checks/systemd/systemd.py

```diff
@@ -0,0 +1,177 @@
+
+
+class SystemdCheck(AgentCheck):
+
+    UNIT_STATUS_SC = 'systemd.unit.active'
+
+    def __init__(self, name, init_config, instances):
+        if instances is not None and len(instances) > 1:
+            raise ConfigurationError('Systemd check only supports one configured instance.')
+        super(SystemdCheck, self).__init__(name, init_config, instances)
+
+        instance = instances[0]
+        self.units_watched = instance.get('units', [])
+        self.tags = instance.get('tags', [])
+        # report state of all systemd units found
+        self.report_status = instance.get('report_status', False)
+
+        # cache to store unit states and compare them at the next check run
+        self.unit_cache = defaultdict(dict)
+
+        # unit_cache = {
+        #   "<unit_name>": "<unit_status>",
+        #   "cron.service": "inactive",
+        #   "ssh.service": "active"
+        # }
+
+    def check(self, instance):
+        current_unit_status = defaultdict(dict)
+        for u in self.units_watched:
+            current_unit_status[u] = self.get_state_single_unit(u)
+        # initialize the cache if it's empty
+        if not self.unit_cache:
+            self.unit_cache = current_unit_status
+
+        for unit in self.units_watched:
+            self.send_service_checks(unit, self.get_state_single_unit(unit), self.tags)
+
+        # find out which units have changed state, units that can no longer be found or units found
+        changed_units, created_units, deleted_units = self.list_status_change(current_unit_status)
+
+        self.report_changed_units(changed_units, self.tags)
+        self.report_deleted_units(deleted_units, self.tags)
+        self.report_created_units(created_units, self.tags)
+
+        if self.report_status:
+            all_units = self.get_all_unit_status()
+            self.report_statuses(all_units, self.tags)
+
+        # we update the cache at the end of the check run
+        self.unit_cache = current_unit_status
+
+    def get_all_unit_status(self):
+        m = pystemd.systemd1.Manager(_autoload=True)
+        # list_unit_files is a list of tuples including the unit names/ids and status
+        list_unit_files = m.Manager.ListUnits()
+
+        unit_status = defaultdict(dict)
+
+        for unit in list_unit_files:
+            unit_short_name = unit[0]  # unit name/id
+            unit_state = unit[3]  # unit state
+            unit_status[unit_short_name] = unit_state
+
+        return unit_status
+
+    def list_status_change(self, current_unit_status):
+        # returns a tuple of dicts for unit changes
+        changed = defaultdict(dict)
+        deleted = defaultdict(dict)
+        created = copy.deepcopy(current_unit_status)
+
+        for unit_short_name, previous_unit_status in iteritems(self.unit_cache):
+            # We remove all previous cached entries from the current unit status
+            created.pop(unit_short_name, None)
+            # del created[unit_short_name]
```

Comment:
Unsure about this one, `del` vs `pop()`

Reply:
Both are fine. `pop` will return the value and here you don't need it tho

Reply:
My `test_cache` is failing with a `KeyError: <unit_id>` if I use `del`